ibm_was/datadog_checks/ibm_was/ibm_was.py

```diff
@@ -0,0 +1,103 @@
+
+from datadog_checks.base import AgentCheck, is_affirmative
+from . import metrics, validation
+
+
+class IbmWasCheck(AgentCheck):
+
+    SERVICE_CHECK_CONNECT = "ibm_was.can_connect"
+
+    def check(self, instance):
+        validation.validate_config(instance)
+        collect_stats = self.setup_configured_stats(instance)
+        nested_tags, metric_categories = self.append_custom_queries(instance, collect_stats)
+
+        custom_tags = instance.get('custom_tags', [])
+
+        data = self.make_request("servers")
+        server_data_xml = etree.fromstring(data)
+        node_list = self.get_node_from_root(server_data_xml, "Node")
+
+        for node in node_list:
+            server_list = self.get_node_from_root(node, 'Server')
+            node_tags = list(custom_tags)
+            node_tags.append('node:{}'.format(node.get('name')))
+            for server in server_list:
+                server_tags = ['server:{}'.format(server.get('name'))]
+                server_tags.extend(node_tags)
+
+                for category, prefix in iteritems(metric_categories):
+                    self.log.debug("Collecting {} stats".format(category))
+                    if collect_stats.get(category):
+                        stats = self.get_node_from_name(server, category)
+                        self.process_stats(stats, prefix, metric_categories, nested_tags, server_tags)
+
+    def get_node_from_name(self, xml_data, path):
+        # XMLPath returns a list, but there should only be one element here since we start
+        # the search within a given Node/server
+        data = xml_data.xpath('//Stat[normalize-space(@name)="{}"]'.format(path))
+        if len(data):
+            return data[0]
+        else:
+            self.warning('Error finding {} stats in XML output.'.format(path))
+            return []
+
+    def get_node_from_root(self, xml_data, path):
+        return xml_data.xpath('//{}'.format(path))
+
+    # The XML will have Stat Nodes and Nodes that contain the metrics themselves
+    # We have to recursively go through each Stat Node to properly setup tags
+    # where each Stat will have a different tag key depending on the context.
+    def process_stats(self, stats, prefix, metric_categories, nested_tags, tags, recursion_level=0):
+        for child in stats:
+            if child.tag in metrics.METRIC_VALUE_FIELDS:
+                self.submit_metrics(child, prefix, tags)
+            elif child.tag in metrics.CATEGORY_FIELDS:
+                recursion_tags = tags + ["{}:{}".format(
+                    nested_tags.get(prefix)[recursion_level], child.get('name')
+                )]
+                self.process_stats(child, prefix, metric_categories, nested_tags, recursion_tags, recursion_level+1)
+
+    def submit_metrics(self, child, prefix, tags):
+        value = child.get(metrics.METRIC_VALUE_FIELDS[child.tag])
+        self.gauge('ibmwas.{}.{}'.format(prefix, child.get('name')), value, tags=tags)
```

Comment:
- Are they always gauges ? Even `CountStatistics` ?
- We should probably call `self.normalize` on the metric name. There might be whitespaces and such in the `name` attribute of the XML element, WDYT ?

Reply:
I have to think on this one. The remainder of the comments should be addressed now.

Reply:
There is a definition of the PMI Data types here - https://www.ibm.com/support/knowledgecenter/en/SSAW57_8.5.5/com.ibm.websphere.nd.multiplatform.doc/ae/rprf_dataclass.html

I think all could be gauges with the exception of the CountStatistics, which could be a rate? What do you think?

Reply:
For CountStatistics, I'd use `monotonic_count`
From this document and the fixture you added, I also see `DoubleStatistics` (for which I would use `rate`) as well as `AverageStatistics` (`gauge`, similarly to `TimeSatistics`). Any reason you did not add them ?